crates/ra_parser/src/grammar/attributes.rs

```diff
@@ -22,8 +22,25 @@ fn attribute(p: &mut Parser, inner: bool) {
         p.bump(T![!]);
     }
 
-    if p.at(T!['[']) {
-        items::token_tree(p);
+    if p.eat(T!['[']) {
+        paths::use_path(p);
+
+        let is_delimiter = |p: &mut Parser| match p.current() {
+            T!['('] | T!['['] | T!['{'] => true,
+            _ => false,
+        };
+
+        if p.eat(T![=]) {
+            if expressions::literal(p).is_none() {
+                p.error("expected literal");
+            }
+        } else if is_delimiter(p) {
+            items::token_tree(p);
+        }
```

Comment:
Nit: in the grammar, we try to avoid high-order functions, as at seems like they don't buy us much in this case.

Let's change this to

```rust
match p.current() {
    T![=] => { p.bump(T![]]); expressions::literal(p)... },
    T!['('] | T!['['] | T!['{'] => items::token_tree(p),
}
```

This way, we also look at the current token only once, instead of twice.